ClassNotFound during static init build step corrupts delayed logging

### Describe the bug

An underlying classnotfound exception during a static init build step is causing the delayed logging handling to fail with a NullPointerException and the delayed logging is incomplete.

### Expected behavior

The delayed logging should work and/or the underlying exception should be propogated to the caller of QuarkusEntryPoint.main().

### Actual behavior

With an underlying class not found exception during a static init build step (captured via a debug session):

```
java.util.ServiceConfigurationError: org.keycloak.authentication.AuthenticatorFactory: Provider org.keycloak.examples.authenticator.SecretQuestionAuthenticatorFactory could not be instantiated
	at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:586)
	at java.base/java.util.ServiceLoader$ProviderImpl.newInstance(ServiceLoader.java:813)
	at java.base/java.util.ServiceLoader$ProviderImpl.get(ServiceLoader.java:729)
	at java.base/java.util.ServiceLoader$3.next(ServiceLoader.java:1403)
	at org.keycloak.provider.DefaultProviderLoader.load(DefaultProviderLoader.java:60)
	at org.keycloak.provider.ProviderManager.load(ProviderManager.java:107)
	at org.keycloak.quarkus.deployment.KeycloakProcessor.loadFactories(KeycloakProcessor.java:705)
	at org.keycloak.quarkus.deployment.KeycloakProcessor.configureKeycloakSessionFactory(KeycloakProcessor.java:468)
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
	at java.base/java.lang.Thread.run(Thread.java:1583)
	at org.jboss.threads.JBossThread.run(JBossThread.java:483)
Caused by: java.lang.NoClassDefFoundError: org/keycloak/examples/authenticator/SecretQuestionAuthenticator
	at org.keycloak.examples.authenticator.SecretQuestionAuthenticatorFactory.<clinit>(SecretQuestionAuthenticatorFactory.java:39)
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized0(Native Method)
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized(Unsafe.java:1160)
	at java.base/jdk.internal.reflect.MethodHandleAccessorFactory.ensureClassInitialized(MethodHandleAccessorFactory.java:300)
	at java.base/jdk.internal.reflect.MethodHandleAccessorFactory.newConstructorAccessor(MethodHandleAccessorFactory.java:103)
	at java.base/jdk.internal.reflect.ReflectionFactory.newConstructorAccessor(ReflectionFactory.java:200)
	at java.base/java.lang.reflect.Constructor.acquireConstructorAccessor(Constructor.java:549)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
	at java.base/java.util.ServiceLoader$ProviderImpl.newInstance(ServiceLoader.java:789)
	... 15 more
Caused by: java.lang.ClassNotFoundException: org.keycloak.examples.authenticator.SecretQuestionAuthenticator
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:535)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:481)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:533)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:481)
	... 25 more
```

There is then an attempt to purge the delayed logging entries:

```
The DelayedHandler was closed before any children handlers were configured. Messages will be written to stderr.
2025-01-16 09:59:07,623 DEBUG [org.jboss.logging] (main) Logging Provider: org.jboss.logging.JBossLogManagerProvider

... more log entries
```

However that triggers a NullPointerException, so that Keycloak shows:

```
java.lang.NullPointerException
	at java.base/java.util.Objects.requireNonNull(Objects.java:233)
	at java.base/java.util.ImmutableCollections$List12.<init>(ImmutableCollections.java:563)
	at java.base/java.util.List.of(List.java:937)
	at io.quarkus.paths.OpenContainerPathTree.getRoots(OpenContainerPathTree.java:96)
	at io.quarkus.paths.SharedArchivePathTree$CallerOpenPathTree.getRoots(SharedArchivePathTree.java:142)
	at io.quarkus.bootstrap.classloading.PathTreeClassPathElement.toString(PathTreeClassPathElement.java:214)
	at java.base/java.util.Formatter$FormatSpecifier.printString(Formatter.java:3158)
	at java.base/java.util.Formatter$FormatSpecifier.print(Formatter.java:3036)
	at java.base/java.util.Formatter.format(Formatter.java:2791)
	at java.base/java.util.Formatter.format(Formatter.java:2728)
	at java.base/java.lang.String.format(String.java:4390)
	at org.jboss.logmanager.ExtFormatter.formatMessagePrintf(ExtFormatter.java:144)
	at org.jboss.logmanager.ExtFormatter.formatMessage(ExtFormatter.java:91)
	at org.jboss.logmanager.formatters.Formatters$16.renderRaw(Formatters.java:832)
	at org.jboss.logmanager.formatters.Formatters$JustifyingFormatStep.render(Formatters.java:227)
	at org.jboss.logmanager.formatters.MultistepFormatter.format(MultistepFormatter.java:90)
	at org.jboss.logmanager.ExtFormatter.format(ExtFormatter.java:58)
	at io.quarkus.bootstrap.logging.QuarkusDelayedHandler.close(QuarkusDelayedHandler.java:157)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:35)
	at org.keycloak.quarkus.runtime.cli.Picocli.build(Picocli.java:945)
	at org.keycloak.quarkus.runtime.cli.command.Build.run(Build.java:91)
	at picocli.CommandLine.executeUserObject(CommandLine.java:2030)
	...
	at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:68)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:62)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:33)
```

### How to Reproduce?

Add a jar to the provider directory with a missing class - I used https://github.com/keycloak/keycloak-quickstarts/tree/main/extension/authenticator and remove the Provider classes referenced by the Factories, see also https://github.com/keycloak/keycloak/issues/36482

Run a build

I can also try to isolate this to a non-Keycloak quarkus project if needed.

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

3.15.2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

The Keycloak user reported that this behavior did not occur with Keycloak based upon Quarkus 3.8.5.

@gsmet with your path applied the delayed logging closure now shows:

```

...

2025-01-22 16:23:34,003 DEBUG [io.quarkus.bootstrap.classloading.QuarkusClassLoader.lifecycle] (main) Closing class loader QuarkusClassLoader:Deployment Class Loader: PROD for keycloak@48f5bde6: java.lang.RuntimeException: Created to log a stacktrace
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.close(QuarkusClassLoader.java:640)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.createProductionApplication(AugmentActionImpl.java:177)
	at io.quarkus.deployment.mutability.ReaugmentTask.main(ReaugmentTask.java:70)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doReaugment(QuarkusEntryPoint.java:90)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:49)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:33)
	at org.keycloak.quarkus.runtime.cli.Picocli.build(Picocli.java:945)
	at org.keycloak.quarkus.runtime.cli.command.Build.run(Build.java:91)
	at picocli.CommandLine.executeUserObject(CommandLine.java:2030)
	at picocli.CommandLine.access$1500(CommandLine.java:148)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2465)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2457)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2419)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2277)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2421)
	at picocli.CommandLine.execute(CommandLine.java:2174)
	at org.keycloak.quarkus.runtime.cli.Picocli.parseAndRun(Picocli.java:130)
	at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:94)
	at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:68)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:62)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:33)


2025-01-22 16:23:34,004 DEBUG [io.quarkus.bootstrap.classloading.QuarkusClassLoader.lifecycle] (main) Closing class loader QuarkusClassLoader:Augmentation Class Loader: PROD for keycloak@73db4768: java.lang.RuntimeException: Created to log a stacktrace
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.close(QuarkusClassLoader.java:640)
	at io.quarkus.bootstrap.app.CuratedApplication.close(CuratedApplication.java:429)
	at io.quarkus.deployment.mutability.ReaugmentTask.main(ReaugmentTask.java:59)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doReaugment(QuarkusEntryPoint.java:90)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:49)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:33)
	at org.keycloak.quarkus.runtime.cli.Picocli.build(Picocli.java:945)
	at org.keycloak.quarkus.runtime.cli.command.Build.run(Build.java:91)
	at picocli.CommandLine.executeUserObject(CommandLine.java:2030)
	at picocli.CommandLine.access$1500(CommandLine.java:148)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2465)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2457)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2419)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2277)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2421)
	at picocli.CommandLine.execute(CommandLine.java:2174)
	at org.keycloak.quarkus.runtime.cli.Picocli.parseAndRun(Picocli.java:130)
	at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:94)
	at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:68)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:62)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:33)


ERROR: Failed to run 'build' command.
ERROR: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step org.keycloak.quarkus.deployment.KeycloakProcessor#configureKeycloakSessionFactory threw an exception: java.util.ServiceConfigurationError: org.keycloak.authentication.AuthenticatorFactory: Provider org.keycloak.examples.authenticator.SecretQuestionAuthenticatorFactory could not be instantiated
	at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:586)
	at java.base/java.util.ServiceLoader$ProviderImpl.newInstance(ServiceLoader.java:813)
	at java.base/java.util.ServiceLoader$ProviderImpl.get(ServiceLoader.java:729)
	at java.base/java.util.ServiceLoader$3.next(ServiceLoader.java:1403)
	at org.keycloak.provider.DefaultProviderLoader.load(DefaultProviderLoader.java:60)
	at org.keycloak.provider.ProviderManager.load(ProviderManager.java:107)
	at org.keycloak.quarkus.deployment.KeycloakProcessor.loadFactories(KeycloakProcessor.java:700)
	at org.keycloak.quarkus.deployment.KeycloakProcessor.configureKeycloakSessionFactory(KeycloakProcessor.java:467)
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
	at java.base/java.lang.Thread.run(Thread.java:1583)
	at org.jboss.threads.JBossThread.run(JBossThread.java:483)
Caused by: java.lang.NoClassDefFoundError: org/keycloak/examples/authenticator/SecretQuestionAuthenticator
	at org.keycloak.examples.authenticator.SecretQuestionAuthenticatorFactory.<clinit>(SecretQuestionAuthenticatorFactory.java:39)
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized0(Native Method)
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized(Unsafe.java:1160)
	at java.base/jdk.internal.reflect.MethodHandleAccessorFactory.ensureClassInitialized(MethodHandleAccessorFactory.java:300)
	at java.base/jdk.internal.reflect.MethodHandleAccessorFactory.newConstructorAccessor(MethodHandleAccessorFactory.java:103)
	at java.base/jdk.internal.reflect.ReflectionFactory.newConstructorAccessor(ReflectionFactory.java:200)
	at java.base/java.lang.reflect.Constructor.acquireConstructorAccessor(Constructor.java:549)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
	at java.base/java.util.ServiceLoader$ProviderImpl.newInstance(ServiceLoader.java:789)
	... 15 more
Caused by: java.lang.ClassNotFoundException: org.keycloak.examples.authenticator.SecretQuestionAuthenticator
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:535)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:481)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:533)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:481)
	... 25 more

ERROR: Build failure: Build failed due to errors
	[error]: Build step org.keycloak.quarkus.deployment.KeycloakProcessor#configureKeycloakSessionFactory threw an exception: java.util.ServiceConfigurationError: org.keycloak.authentication.AuthenticatorFactory: Provider org.keycloak.examples.authenticator.SecretQuestionAuthenticatorFactory could not be instantiated
	at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:586)
	at java.base/java.util.ServiceLoader$ProviderImpl.newInstance(ServiceLoader.java:813)
	at java.base/java.util.ServiceLoader$ProviderImpl.get(ServiceLoader.java:729)
	at java.base/java.util.ServiceLoader$3.next(ServiceLoader.java:1403)
	at org.keycloak.provider.DefaultProviderLoader.load(DefaultProviderLoader.java:60)
	at org.keycloak.provider.ProviderManager.load(ProviderManager.java:107)
	at org.keycloak.quarkus.deployment.KeycloakProcessor.loadFactories(KeycloakProcessor.java:700)
	at org.keycloak.quarkus.deployment.KeycloakProcessor.configureKeycloakSessionFactory(KeycloakProcessor.java:467)
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
	at java.base/java.lang.Thread.run(Thread.java:1583)
	at org.jboss.threads.JBossThread.run(JBossThread.java:483)
Caused by: java.lang.NoClassDefFoundError: org/keycloak/examples/authenticator/SecretQuestionAuthenticator
	at org.keycloak.examples.authenticator.SecretQuestionAuthenticatorFactory.<clinit>(SecretQuestionAuthenticatorFactory.java:39)
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized0(Native Method)
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized(Unsafe.java:1160)
	at java.base/jdk.internal.reflect.MethodHandleAccessorFactory.ensureClassInitialized(MethodHandleAccessorFactory.java:300)
	at java.base/jdk.internal.reflect.MethodHandleAccessorFactory.newConstructorAccessor(MethodHandleAccessorFactory.java:103)
	at java.base/jdk.internal.reflect.ReflectionFactory.newConstructorAccessor(ReflectionFactory.java:200)
	at java.base/java.lang.reflect.Constructor.acquireConstructorAccessor(Constructor.java:549)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
	at java.base/java.util.ServiceLoader$ProviderImpl.newInstance(ServiceLoader.java:789)
	... 15 more
Caused by: java.lang.ClassNotFoundException: org.keycloak.examples.authenticator.SecretQuestionAuthenticator
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:535)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:481)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:533)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:481)
	... 25 more
```

Then keycloak emits an error related to the underlying problem because we're catching the actual exception:

```
ERROR: org.keycloak.authentication.AuthenticatorFactory: Provider org.keycloak.examples.authenticator.SecretQuestionAuthenticatorFactory could not be instantiated
ERROR: org/keycloak/examples/authenticator/SecretQuestionAuthenticator
ERROR: org.keycloak.examples.authenticator.SecretQuestionAuthenticator
```

This seems good to me.

--- a/independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/ArchivePathTree.java
+++ b/independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/ArchivePathTree.java
@@ -229,6 +229,11 @@ public class ArchivePathTree extends PathTreeWithManifest implements PathTree {
                 && manifestEnabled == other.manifestEnabled;
     }
 
+    @Override
+    public String toString() {
+        return archive.toString();
+    }
+
     protected class OpenArchivePathTree extends OpenContainerPathTree {
 
         private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
@@ -402,6 +407,11 @@ public class ArchivePathTree extends PathTreeWithManifest implements PathTree {
         @Override
         public PathTree getOriginalTree() {
             return ArchivePathTree.this;
+        }
+
+        @Override
+        public String toString() {
+            return ArchivePathTree.this.toString();
         }
     }
 }

--- a/independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/EmptyPathTree.java
+++ b/independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/EmptyPathTree.java
@@ -76,4 +76,9 @@ public class EmptyPathTree implements OpenPathTree {
     public PathTree getOriginalTree() {
         return this;
     }
+
+    @Override
+    public String toString() {
+        return "<empty>";
+    }
 }

--- a/independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/FilePathTree.java
+++ b/independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/FilePathTree.java
@@ -137,4 +137,13 @@ class FilePathTree implements OpenPathTree {
         FilePathTree other = (FilePathTree) obj;
         return Objects.equals(file, other.file) && Objects.equals(pathFilter, other.pathFilter);
     }
+
+    @Override
+    public String toString() {
+        if (pathFilter == null) {
+            return file.toString();
+        }
+
+        return file + " (filtered)";
+    }
 }

--- a/independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/FilteredPathTree.java
+++ b/independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/FilteredPathTree.java
@@ -89,6 +89,11 @@ public class FilteredPathTree implements PathTree {
         return new OpenFilteredPathTree(original.open(), filter);
     }
 
+    @Override
+    public String toString() {
+        return original.toString() + " (filtered)";
+    }
+
     private static class OpenFilteredPathTree extends FilteredPathTree implements OpenPathTree {
 
         private final OpenPathTree original;

--- a/independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/MultiRootPathTree.java
+++ b/independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/MultiRootPathTree.java
@@ -9,6 +9,7 @@ import java.util.List;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.function.Consumer;
 import java.util.function.Function;
+import java.util.stream.Collectors;
 
 public class MultiRootPathTree implements OpenPathTree {
 
@@ -191,5 +192,10 @@ public class MultiRootPathTree implements OpenPathTree {
             return false;
         MultiRootPathTree other = (MultiRootPathTree) obj;
         return Arrays.equals(trees, other.trees);
+    }
+
+    @Override
+    public String toString() {
+        return roots.stream().map(p -> p.toString()).collect(Collectors.joining(", ", "[", "]"));
     }
 }

--- a/independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/OpenContainerPathTree.java
+++ b/independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/OpenContainerPathTree.java
@@ -186,4 +186,9 @@ public abstract class OpenContainerPathTree extends PathTreeWithManifest impleme
                 && Objects.equals(pathFilter, other.pathFilter)
                 && manifestEnabled == other.manifestEnabled;
     }
+
+    @Override
+    public String toString() {
+        return getContainerPath().toString();
+    }
 }

--- a/independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/SharedArchivePathTree.java
+++ b/independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/SharedArchivePathTree.java
@@ -57,7 +57,7 @@ class SharedArchivePathTree extends ArchivePathTree {
             return new CallerOpenPathTree(lastOpen);
         }
         try {
-            this.lastOpen = new SharedOpenArchivePathTree(archive, openFs());
+            this.lastOpen = new SharedOpenArchivePathTree(openFs());
         } catch (IOException e) {
             throw new UncheckedIOException(e);
         }
@@ -68,7 +68,7 @@ class SharedArchivePathTree extends ArchivePathTree {
 
         private final AtomicInteger users = new AtomicInteger(1);
 
-        protected SharedOpenArchivePathTree(Path archivePath, FileSystem fs) {
+        protected SharedOpenArchivePathTree(FileSystem fs) {
             super(fs);
             openCount.incrementAndGet();
         }
@@ -105,6 +105,11 @@ class SharedArchivePathTree extends ArchivePathTree {
             } finally {
                 writeLock().unlock();
             }
+        }
+
+        @Override
+        public String toString() {
+            return SharedArchivePathTree.this.toString();
         }
     }
 
@@ -201,6 +206,11 @@ class SharedArchivePathTree extends ArchivePathTree {
             } finally {
                 delegate.writeLock().unlock();
             }
+        }
+
+        @Override
+        public String toString() {
+            return delegate.toString();
         }
     }
 }

--- a/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/PathTreeClassPathElement.java
+++ b/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/PathTreeClassPathElement.java
@@ -14,7 +14,6 @@ import java.security.ProtectionDomain;
 import java.security.cert.Certificate;
 import java.util.ArrayList;
 import java.util.HashSet;
-import java.util.Iterator;
 import java.util.List;
 import java.util.Objects;
 import java.util.Set;
@@ -216,13 +215,7 @@ public class PathTreeClassPathElement extends AbstractClassPathElement {
         if (getDependencyKey() != null) {
             sb.append(getDependencyKey().toGacString()).append(" ");
         }
-        final Iterator<Path> i = pathTree.getRoots().iterator();
-        if (i.hasNext()) {
-            sb.append(i.next());
-            while (i.hasNext()) {
-                sb.append(",").append(i.next());
-            }
-        }
+        sb.append(pathTree.getOriginalTree().toString());
         sb.append(" runtime=").append(isRuntime());
         final Set<String> resources = this.resources;
         sb.append(" resources=").append(resources == null ? "null" : resources.size());

